File: src/compressor.cpp
#include "compressor.hpp"

double Compressor::db_to_amp(double db)
{
	return (pow(10, db / 20.0f) * sample_range);
}

double Compressor::amp_to_db(double amp)
{
	return 20.0f * log10(amp / sample_range);
}

Compressor::Compressor()
{
	threshold_amp = db_to_amp(threshold - 3.162); //these values are being adjusted for compression purposes
	normalize_amp = db_to_amp(normalize_db);
	noise_floor_amp = db_to_amp(noise_floor - 3.162);
}

//getters and setters
void Compressor::set_threshold(double threshold)
{
	this->threshold = threshold;
	threshold_amp = db_to_amp(threshold);
}
double Compressor::get_threshold()
{
	return threshold;
}

void Compressor::set_normalize(double normalize)
{
	this->normalize_db = normalize;
	normalize_amp = db_to_amp(normalize_db);
}
double Compressor::get_normalize()
{
	return normalize_db;
}

void Compressor::set_noise_floor(double noise_floor)
{
	this->noise_floor = noise_floor;
	noise_floor_amp = db_to_amp(noise_floor);
}
double Compressor::get_noise_floor()
{
	return noise_floor;
}

void Compressor::set_ratio(double ratio)
{
	this->ratio = ratio;
}
double Compressor::get_ratio()
{
	return ratio;
}

void Compressor::set_attack_time_ms(int attack)
{
	attack_time = attack;
}
int Compressor::get_attack_time_ms()
{
	return attack_time;
}

void Compressor::set_release_time_ms(int release)
{
	release_time = release;
}
int Compressor::get_release_time_ms()
{
	return release_time;
}

void Compressor::set_sample_window_ms(int window)
{
	sample_window = window;
}
int Compressor::get_sample_window_ms()
{
	return sample_window;
}

void Compressor::compress(Wav& wav)
{
	//initialize things that need data from the wav file
	samples_per_sec = wav.samples_per_sec;
	samples_in_window = samples_per_sec * (double)(sample_window / 1000.0f);
	attack_samples = samples_per_sec * (double)(attack_time / 1000.0f);
	release_samples = samples_per_sec * (double)(release_time / 1000.0f);

	double gain_adjust = 0;
	double output_gain = 0;
	double attack_step = 1 / (double)attack_samples;
	double release_step = 1 / (double)release_samples;

	//this queue holds original values
	//since we are modifying in-place, we need a copy of them
	queue<int64_t> old_values;

	uint64_t total = 0;
	int16_t* data = (int16_t*)((void*)wav.data.data());

	//iterate through whole file
	for(int i = 0; i < wav.samples; i++)
	{
		uint64_t cur = pow(data[i], 2);
		total += cur;
		old_values.push(cur);

		//remove old sample from window if we go above the window size
		if(old_values.size() == (samples_in_window + 1))
		{
			uint64_t prev = old_values.front();
			total -= prev;
			old_values.pop();
		}

		//calculate RMS values (rolling)
		double RMS = sqrt(total / old_values.size());

		//attack (lower volume)
		if(RMS > threshold_amp)
		{
			//amount (in decibels) we wish to be over the threshold
			double overamp = ((amp_to_db(RMS) - 3.162) - threshold) / ratio;

			//target amplitude (converted back from db)
			double target_amp = db_to_amp(threshold + overamp - 3.162);

			//difference in amplitude between our average and our target
			double diff = RMS - target_amp;

			if(gain_adjust <= 1)
			{
				gain_adjust += attack_step;
			}

			//calculate output gain adjustment
			output_gain = (RMS - (diff * gain_adjust)) / RMS;
			data[i] *= output_gain;
		}

		//release(raise volume back to normal)
		else
		{
			double diff = 1 - output_gain;

			if(gain_adjust >= 0)
			{
				gain_adjust -= release_step;
			}

			output_gain = 1 - (diff * gain_adjust);
			data[i] *= output_gain;
		}
	}
}

void Compressor::remove_peaks(Wav& wav)
{

}

void Compressor::normalize(Wav& wav)
{
	//the sample size has been set to 50ms to simulate what it would be like in a real streaming setup
	int sample_size = (50 / (double)1000) * wav.samples_per_sec;
	int16_t* data = (int16_t*)(void*)wav.data.data();

	//peak value is over the entire file
	int peak = 0;
	for(int i = 0; i < wav.samples; i += sample_size)
	{
		//length is the sample size unless there are not enough samples remaining
		int len = sample_size;
		if(wav.samples - i < sample_size)
		{
			len = wav.samples - i;
		}

		//check for any value that exceeds the peak
		for(int x = i; x < i+len; x++)
		{
			int cur = abs(data[x]);
			if(cur > peak)
			{
				peak = cur;
			}
		}

		//calculate the amount of gain to add or subtract
		double gain = normalize_amp / peak;

		//apply gain to entire window
		for(int x = i; x < i+len; x++)
		{
			data[x] *= gain;
		}
	}

}

File: prototype/compressor.py
import wave
import math
import numpy as np

#compress the dynamic range of a wav file (as a proof of concept)
class Compressor:

	#we will be checking that every wav file is signed 16 bit PCM
	__sample_range: int = 2**15

	#any value above this will be decreased by the downward compressor(db)
	threshold: float = -18
	__threshold_amp: int = 0

	#what value to normalize the audio to after it has been compressed
	normalize_db: float = -6
	__normalize_amp: int = 0

	#ignore all samples under this value
	noise_floor: float = -40
	__noise_floor_amp: int = 0

	#how aggressive do we want the compressor to be
	ratio: float = 4

	#attack and release params
	attack_time_ms: int = 20
	__attack_time_samples: int = 0

	release_time_ms: int = 500
	__release_time_samples: int = 0

	#sample window params
	sample_window_ms: int = 50

	#convert db to amplitude(16 bit signed)
	def __db_to_amp(self, db):
		return 10**(db / 20) * self.__sample_range

	#convert amplitude to db(16 bit signed)
	def __amp_to_db(self, amp):
		return 20 * math.log10(amp / self.__sample_range)

	#compress an audio file
	def compress(self, in_file: str, out_file: str):

		#open the wave file and establish some data
		inf = wave.open(in_file, "rb")
		wavparams = inf.getparams()

		#check to make sure our data is what we expect
		if(wavparams.sampwidth != 2):
			print("Compressor Error: wav file does not contain 16 bit signed PCM data.")
			return

		#read all samples into a buffer
		dt = np.dtype(np.int16)
		dt = dt.newbyteorder("<")
		samples = np.array(np.frombuffer(inf.readframes(wavparams.nframes), dtype=dt))
		samples.setflags(write=True)
		sample_rate = wavparams.framerate
		inf.close()

		#initialize amp values
		self.__threshold_amp = self.__db_to_amp(self.threshold)

		#initialize sample values
		self.__attack_time_samples = int((sample_rate / 1000) * self.attack_time_ms)
		self.__release_time_samples = int((sample_rate / 1000) * self.release_time_ms)
		self.__noise_floor_amp = self.__db_to_amp(self.noise_floor)

		#50ms window size
		window_size: int = int(sample_rate * (self.sample_window_ms / 1000))
		gain_adjust: float = 0
		output_gain: float = 1
		attack_step: float = 1 / self.__attack_time_samples
		release_step: float = 1 / self.__release_time_samples
		
		#iterate through the samples by byte
		for i in range(0, len(samples), window_size):
			window = samples[i:i+window_size]

			#calculate amplitude of a sample (RMS)
			data = np.sqrt(np.mean(window.astype(np.int32)**2))
			
			#attack adjustment (lower volume)
			if(data > self.__threshold_amp):
				overamp = (data - self.__threshold_amp) / self.ratio
				target_amp = (self.__threshold_amp + overamp)
				diff = data - target_amp

				for x in range(0, len(window)):
					
					#change attack and release values accordingly
					if(gain_adjust < 1):
						gain_adjust += attack_step
					
					#adjust the value
					output_gain = (data - (diff * gain_adjust)) / data
					window[x] *= output_gain

			#release adjustment (raise volume)
			else:
				for x in range(0, len(window), 2):

					#difference between full gain and current gain
					diff = 1 - output_gain
					
					#change attack and release values accordingly
					if(gain_adjust > 0):
						gain_adjust -= release_step
					
					#this calculation will go up from output gain to full volume
					release_output = 1 - (diff * gain_adjust)
					window[x] *= release_output
		
		outf = wave.open(out_file, "wb")
		outf.setparams(wavparams)
		outf.writeframes(samples.tobytes())
		outf.close()
